onadata/libs/utils/xml.py

```diff
@@ -0,0 +1,183 @@
+        bind_node = self._root_node.createElement('bind')
+        bind_node.setAttribute(
+            'nodeset', f'/{self._unique_id}/_{self._unique_id}__disclaimer'
+        )
+        bind_node.setAttribute('readonly', 'true()')
+        bind_node.setAttribute('required', 'false()')
+        bind_node.setAttribute('type', 'string')
+        bind_node.setAttribute('relevant', 'false()')
+        model_node.appendChild(bind_node)
+
+        # Inject note node inside <{self._unique_id}>
+        instance_node = model_node.getElementsByTagName('instance')[0]
+        instance_node = instance_node.getElementsByTagName(self._unique_id)[0]
+        instance_node.appendChild(
+            self._root_node.createElement(f'_{self._unique_id}__disclaimer')
+        )
+
+    def _add_disclaimer_input(
+        self,
+        translated: bool,
+        disclaimers_dict: dict,
+        default_language_code: str,
+    ):
+        """
+
+        """
+        body_node = self._root_node.getElementsByTagName('h:body')[0]
+        disclaimer_input = self._root_node.createElement('input')
+        disclaimer_input_label = self._root_node.createElement('label')
+        disclaimer_input.setAttribute('appearance', 'kobo-disclaimer')
+        disclaimer_input.setAttribute(
+            'ref', f'/{self._unique_id}/_{self._unique_id}__disclaimer'
+        )
+
+        if translated:
+            itext = f'/{self._unique_id}/_{self._unique_id}__disclaimer:label'
+            disclaimer_input_label.setAttribute(
+                'ref',
+                f"jr:itext('{itext}')",
+            )
+        else:
+            disclaimer_input_label.appendChild(
+                self._root_node.createTextNode(
+                    disclaimers_dict[default_language_code]
+                )
+            )
+
+        disclaimer_input.appendChild(disclaimer_input_label)
+        body_node.appendChild(disclaimer_input)
+
+    def _add_translation_nodes(
+        self, disclaimers_dict: dict, default_language_code: str
+    ):
+        """
+        Add <itext> nodes to <instance> if translations are detected.
+        Will add only translations that match form translations.
+        """
+
+        translation_nodes = []
+        languages = []
+        for n in self._root_node.getElementsByTagName('itext')[0].childNodes:
+            if n.nodeType == Node.ELEMENT_NODE and n.tagName == 'translation':
+                languages.append(n.getAttribute('lang'))
+                translation_nodes.append(n)
+                disclaimer_translation = self._root_node.createElement('text')
+                disclaimer_translation.setAttribute(
+                    'id',
+                    f'/{self._unique_id}/_{self._unique_id}__disclaimer:label',
+                )
+                value = self._root_node.createElement('value')
+                language = n.getAttribute('lang').lower().strip()
+                if m := re.match(r'[^\(]*\(([a-z]{2,})\)', language):
```

Comment:
~~i think some language codes have more characters, like `zh-Hans`~~
not true because of the language lists used by kpi